### PR TITLE
[GEP-18] Revendor `gardener/gardener@v1.45`, cleanup token projection artefacts, adapt to new secrets manager

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area security
/kind enhancement

**What this PR does / why we need it**:
- Vendors g/g and adapts to all breaking changes introduced by https://github.com/gardener/gardener/pull/5803
- Cleans up a lot of migration code around moving to short-lived serviceaccount tokens
- Adapts this extension to the new secrets manager, which makes it ready for supporting the shoot CA rotation feature

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/3292

**Special notes for your reviewer**:
/invite @timebertt 

**Release note**:

```breaking operator
This extension is only compatible with Gardener versions `>= v1.37`.
```
```feature operator
This extension is prepared to support the Shoot CA rotation feature ([GEP-18](https://github.com/gardener/gardener/issues/3292)).
```

### CODE REVIEW

File: pkg/controller/lifecycle/actuator.go [hunk 1]
```diff
@@ -130,29 +133,56 @@ func (a *actuator) Reconcile(ctx context.Context, ex *extensionsv1alpha1.Extensi
...
+				Config: &secretutils.CertificateSecretConfig{
+					Name:       caName,
+					CommonName: caName,
+					CertType:   secretutils.CACert,
+				},
+				Options: []secretsmanager.GenerateOption{secretsmanager.Persist()},
+			},
+			{
+				Config: &secretutils.CertificateSecretConfig{
+					Name:       constants.WebhookTLSecretName,
```
Comment (timebertt) [inline pkg/controller/lifecycle/actuator.go:150]: Typo: `WebhookTLSSecretName`

File: pkg/controller/lifecycle/actuator.go [hunk 2]
```diff
@@ -130,29 +133,56 @@ func (a *actuator) Reconcile(ctx context.Context, ex *extensionsv1alpha1.Extensi
...
+					CertType:   secretutils.ServerCert,
+				},
+				// use current CA for signing server cert to prevent mismatches when dropping the old CA from the webhook
+				// config in phase Completing
+				Options: []secretsmanager.GenerateOption{secretsmanager.SignedByCA(caName, secretsmanager.UseCurrentCA)},
+			},
+		}
+	)
+
+	secretsManager, err := extensionssecretsmanager.SecretsManagerForCluster(ctx, a.logger.WithName("secretsmanager"), clock.RealClock{}, a.client, cluster, constants.SecretsManagerIdentity, nil)
```
Comment (timebertt) [inline pkg/controller/lifecycle/actuator.go:162]: You need to pass the `secretConfigs` in order for the CA to be rotated.
```suggestion
	secretsManager, err := extensionssecretsmanager.SecretsManagerForCluster(ctx, a.logger.WithName("secretsmanager"), clock.RealClock{}, a.client, cluster, constants.SecretsManagerIdentity, secretConfigs)
```

**Review by timebertt — COMMENTED** [review]
From a joint review with @kris94:
One bug and one nit :)

File: pkg/controller/lifecycle/actuator.go (hunk 1 above)
Comment (dimityrmirchev) [reply pkg/controller/lifecycle/actuator.go:150]: Thanks! I can rename this in a separate PR so we can keep the changes to a minimum for this one

Comment (rfranzke) [reply pkg/controller/lifecycle/actuator.go:150]: No problem, it's rather straight-forward now, so I add it to my "address PR review" commit

File: pkg/controller/lifecycle/actuator.go (hunk 2 above)
Comment (rfranzke) [reply pkg/controller/lifecycle/actuator.go:162]: nice catch, thank you!

**Review by timebertt — APPROVED** [review]
Nice!
/lgtm

File: pkg/controller/lifecycle/actuator.go [hunk 3]
```diff
@@ -270,69 +320,13 @@ func (a *actuator) InjectScheme(scheme *runtime.Scheme) error {
...
-}
-
 func getLabels() map[string]string {
 	return map[string]string{
 		"app.kubernetes.io/name": constants.ApplicationName,
 	}
 }
 
-func getSeedResources(oidcReplicas *int32, hibernated bool, namespace, genericKubeconfigName string, shootAccessSecret *gutil.ShootAccessSecret) (map[string][]byte, error) {
+func getSeedResources(oidcReplicas *int32, hibernated bool, namespace, genericKubeconfigName, shootAccessSecretName, serverTLSSecretName string) (map[string][]byte, error) {
```
Comment (dimityrmirchev) [inline pkg/controller/lifecycle/actuator.go:329]: In this function we should probably pass the new `bundle.crt` and use it in the kubeconfig that the apiserver uses to talk to the OWA here
https://github.com/gardener/gardener-extension-shoot-oidc-service/blob/31612df0845d00e3ca1bb252885b080db81288fe/pkg/controller/lifecycle/actuator.go#L345

**Review by dimityrmirchev — CHANGES_REQUESTED** [review]
Thank you for the PR! At first glance - one question and some minor bugs.

File: pkg/controller/lifecycle/actuator.go [hunk 4]
```diff
@@ -130,29 +133,56 @@ func (a *actuator) Reconcile(ctx context.Context, ex *extensionsv1alpha1.Extensi
 		return err
 	}
 
-	genericKubeconfigName := extensions.GenericTokenKubeconfigSecretNameFromCluster(cluster)
-	seedResources, err := getSeedResources(oidcReplicas, hibernated, namespace, genericKubeconfigName, oidcShootAccessSecret)
+	// initialize SecretsManager based on Cluster object
+	var (
+		caName        = "ca-extension-shoot-oidc-service"
```
[the remark this answers is not in the document]
Comment (rfranzke) [reply pkg/controller/lifecycle/actuator.go:138]: No, it's a new, dedicated CA for the extension as described in https://github.com/gardener/gardener/blob/master/docs/extensions/ca-rotation.md#requirements-for-shoot-cluster-ca-rotation.

File: pkg/controller/lifecycle/actuator.go [hunk 5]
```diff
@@ -130,29 +133,56 @@ func (a *actuator) Reconcile(ctx context.Context, ex *extensionsv1alpha1.Extensi
...
+		oidcShootAccessSecret.Secret.Name,
+		secrets[constants.WebhookTLSSecretName].Name,
+	)
 	if err != nil {
 		return err
 	}
 
 	shootResources, err := getShootResources(
-		tlsSecret.Data[secrets.DataKeyCertificateCA],
+		secrets[caName].Data[secretutils.DataKeyCertificateBundle],
```
[the remark this answers is not in the document]
Comment (rfranzke) [reply pkg/controller/lifecycle/actuator.go:185]: Nice catch, thanks!

File: pkg/controller/lifecycle/actuator.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (dimityrmirchev) [reply pkg/controller/lifecycle/actuator.go:138]: Thanks. So If I get this correctly "shoot CA rotation" means that all of these CAs will be rotated. Is that right?

File: pkg/controller/lifecycle/actuator.go (hunk 3 above)
Comment (rfranzke) [reply pkg/controller/lifecycle/actuator.go:329]: OK, got it, thanks!

File: pkg/controller/lifecycle/actuator.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (timebertt) [reply pkg/controller/lifecycle/actuator.go:138]: Yes, exactly. We want to have dedicated CAs, but rotate all of them in lock-step :)

File: pkg/controller/lifecycle/actuator.go [hunk 6]
```diff
@@ -134,44 +134,19 @@ func (a *actuator) Reconcile(ctx context.Context, ex *extensionsv1alpha1.Extensi
...
-				},
-				// use current CA for signing server cert to prevent mismatches when dropping the old CA from the webhook
-				// config in phase Completing
-				Options: []secretsmanager.GenerateOption{secretsmanager.SignedByCA(caName, secretsmanager.UseCurrentCA)},
-			},
-		}
-	)
-
-	secretsManager, err := extensionssecretsmanager.SecretsManagerForCluster(ctx, a.logger.WithName("secretsmanager"), clock.RealClock{}, a.client, cluster, constants.SecretsManagerIdentity, secretConfigs)
+	secretsManager, err := extensionssecretsmanager.SecretsManagerForCluster(ctx, a.logger.WithName("secretsmanager"), clock.RealClock{}, a.client, cluster, secrets.ManagerIdentity, secrets.ConfigsFor(namespace))
```
Comment (timebertt) [inline pkg/controller/lifecycle/actuator.go:137]: Extract a variable for `secrets.ConfigsFor(namespace)` instead of calling it twice?

File: pkg/webhook/kapiserver/ensurer.go [hunk 7]
```diff
@@ -54,29 +63,38 @@ func (e *ensurer) EnsureKubeAPIServerDeployment(ctx context.Context, _ gcontext.
...
 		if err != nil {
 			return err
 		}
+
 		if controller.IsHibernated(cluster) {
 			return nil
 		}
 
-		ensureKubeAPIServerIsMutated(ps, c)
+		secretsManager, err := NewSecretsManager(ctx, e.logger.WithName("secretsmanager"), clock.RealClock{}, e.client, cluster, secrets.ManagerIdentity, secrets.ConfigsFor(new.Namespace))
```
Comment (timebertt) [inline pkg/webhook/kapiserver/ensurer.go:83]: Same here

**Review by timebertt — COMMENTED** [review]
Nits, but rest looks good

**Review by timebertt — DISMISSED** [review]
/lgtm

File: pkg/webhook/kapiserver/ensurer.go [hunk 8]
```diff
@@ -163,3 +162,58 @@ func ensureKubeAPIServerIsMutated(ps *corev1.PodSpec, c *corev1.Container, caBun
 		},
 	})
 }
+
+// getLatestIssuedCABundleSecret returns the oidc-webhook latest CA bundle secret
+func getLatestIssuedCABundleSecret(ctx context.Context, c client.Client, namespace string) (*corev1.Secret, error) {
```
Comment (dimityrmirchev) [inline pkg/webhook/kapiserver/ensurer.go:167]: @rfranzke Can you please review the logic in this function? Just to confirm that this is the desired way of getting the newest ca bundle. Everything else looks good to me.

File: pkg/webhook/kapiserver/ensurer.go [hunk 9]
```diff
@@ -163,3 +162,58 @@ func ensureKubeAPIServerIsMutated(ps *corev1.PodSpec, c *corev1.Container, caBun
...
+	initialIssuedAt, ok := secrets[0].Labels[secretsmanager.LabelKeyIssuedAtTime]
+	if !ok {
+		return nil, &NoIssuedAtTimeError{secretName: secrets[0].Name, namespace: secrets[0].Namespace}
+	}
+
+	initialIssuedAtUnix, err := strconv.ParseInt(initialIssuedAt, 10, 64)
+	if err != nil {
+		return nil, err
+	}
+	currentIssuedAtTime := time.Unix(initialIssuedAtUnix, 0).UTC()
```
Comment (rfranzke) [inline pkg/webhook/kapiserver/ensurer.go:195]: Why is this needed? The loop below has the same code?

Comment (dimityrmirchev) [reply pkg/webhook/kapiserver/ensurer.go:195]: This initializes the initial "max" element of the slice that will be used for comparison in the loop below. I guess that there are some other ways to write this but I thought that this will be easy enough to read.

Comment (rfranzke) [reply pkg/webhook/kapiserver/ensurer.go:195]: OK, I don't request changes since it's your component/you are the owner, I was just wondering. Otherwise, this looks good to me.

Comment (dimityrmirchev) [reply pkg/webhook/kapiserver/ensurer.go:195]: I will extract it into a function so that it is not repeated

Comment (rfranzke) [reply pkg/webhook/kapiserver/ensurer.go:195]: You could simply initialize variables and during the loop check if they are still unitialized. If yes, then the current secret can be stored (similar to https://github.com/gardener/gardener/blob/242a08c7eab6bd310b58d5031df8307fc8f96a6e/pkg/utils/secrets/manager/generate.go#L332-L342)

**Review by dimityrmirchev — APPROVED** [review]
/lgtm
